data-analytics/dataflow-python-examples/dataflow_python_examples/data_generation_for_benchmarking.py

```diff
@@ -0,0 +1,485 @@
+        """
+        This function ensures that the data is all of types that BigQuery expects.
+
+        Args:
+            record: A dictionary generated by faker_schema representing the candidate for a fake
+                    row in our BigQuery table
+            fieldname: name of field we are checking with this call.
+        """
+        # Create a Faker instance for individual parameterized random generation
+        # (ie. minimum date).
+        faker = Faker()
+
+        field = self.get_field_dict(fieldname)
+        # Below handles if the datatype got changed by the faker provider
+        if field[u'type'] == 'STRING' or field[u'type'].find('TIME') > -1:
+            record[fieldname] = unicode(record[fieldname])
+        elif field[u'type'].find('DATE') > -1:
+            # This implements the minimum/maximum date functionality
+            # and avoids regenerating a random date if already obeys min/max date.
+            if record[fieldname] < self.data_gen.min_date \
+                    or record[fieldname] > self.data_gen.max_date:
+                record[fieldname] = faker.date_between(self.data_gen.min_date,
+                                                       self.data_gen.max_date)
+            record[fieldname] = unicode(record[fieldname].strftime('%Y-%m-%d'))
+        elif field[u'type'] == 'INTEGER':
+            # This implements max and sign constraints
+            # and avoids regenerating a random integer if already obeys min/max integer.
+            if record[fieldname] > self.data_gen.max_int:
+                record[fieldname] = np.random.randint(0 if self.data_gen.only_pos
+                                                      else -1 * self.data_gen.max_int,
+                                                      self.data_gen.max_int)
+            if self.data_gen.only_pos and record[fieldname] < 0:
+                record[fieldname] = abs(record[fieldname])
+            record[fieldname] = int(record[fieldname])
+        elif field[u'type'] == 'FLOAT':
+            # This implements max and sign constraints
+            record[fieldname] = faker.pyfloat(math.log10(self.data_gen.max_float),
+                                              self.data_gen.float_precision,
+                                              self.data_gen.only_pos)
+            record[fieldname] = float(record[fieldname])
+
+        # Make some values null based on null_prob.
+        if field[u'mode'] == 'NULLABLE':
+            record[fieldname] = np.random.choice([None, record[fieldname]],
+                                                 p=[self.data_gen.null_prob,
+                                                    1.0 - self.data_gen.null_prob])
+
+        # Pick key at random from foreign keys.
+        # Draw key column from [0, n_keys) if has _key in the name.
+        # This forces key column to no contain nulls
+        if ('_key' in field[u'name'].lower()) or ('_id' in field[u'name'].lower()):
```

Comment:
The outermost parens here are not required

Reply:
Removing unnecessary parens.